rust/src/operations/transaction/conflict_checker.rs

```diff
@@ -0,0 +1,862 @@
+        read_snapshot: &'a DeltaTableState,
+        read_predicates: Option<String>,
+        actions: &'a Vec<Action>,
+        read_whole_table: bool,
+    ) -> DeltaResult<Self> {
+        Ok(Self {
+            txn_id: "".into(),
+            read_predicates,
+            read_app_ids: Default::default(),
+            actions,
+            read_snapshot,
+            read_whole_table,
+        })
+    }
+
+    /// Whether the transaction changed the tables metadatas
+    pub fn metadata_changed(&self) -> bool {
+        self.actions
+            .iter()
+            .any(|a| matches!(a, Action::metaData(_)))
+    }
+
+    #[cfg(feature = "datafusion")]
+    /// Files read by the transaction
+    pub fn read_files(&self) -> Result<impl Iterator<Item = &Add>, CommitConflictError> {
+        if let Some(predicate) = &self.read_predicates {
+            Ok(Either::Left(
+                self.read_snapshot
+                    .files_matching_predicate(&[predicate.clone()])
+                    .map_err(|err| CommitConflictError::Predicate {
+                        source: Box::new(err),
+                    })?,
+            ))
+        } else {
+            Ok(Either::Right(std::iter::empty()))
+        }
+    }
+
+    #[cfg(not(feature = "datafusion"))]
+    /// Files read by the transaction
+    pub fn read_files(&self) -> Result<impl Iterator<Item = &Add>, CommitConflictError> {
+        Ok(self.read_snapshot.files().iter())
+    }
+
+    /// Whether the whole table was read during the transaction
+    pub fn read_whole_table(&self) -> bool {
+        self.read_whole_table
+    }
+}
+
+/// Summary of the Winning commit against which we want to check the conflict
+pub(crate) struct WinningCommitSummary {
+    pub actions: Vec<Action>,
+    pub commit_info: Option<CommitInfo>,
+}
+
+impl WinningCommitSummary {
+    pub async fn try_new(
+        object_store: &dyn ObjectStore,
+        read_version: DeltaDataTypeVersion,
+        winning_commit_version: DeltaDataTypeVersion,
+    ) -> DeltaResult<Self> {
+        let mut actions = Vec::new();
+        let mut version_to_read = read_version + 1;
+
+        while version_to_read <= winning_commit_version {
+            let commit_uri = commit_uri_from_version(winning_commit_version);
+            let commit_log_bytes = object_store.get(&commit_uri).await?.bytes().await?;
+
+            let reader = BufReader::new(Cursor::new(commit_log_bytes));
+            for maybe_line in reader.lines() {
+                let line = maybe_line?;
+                actions.push(serde_json::from_str::<Action>(line.as_str()).map_err(|e| {
+                    DeltaTableError::InvalidJsonLog {
+                        json_err: e,
+                        version: winning_commit_version,
+                        line,
+                    }
+                })?);
+            }
+            version_to_read += 1;
+        }
+
+        // TODO how to handle commit info for multiple read commits?
```

Comment:
Yeah, if there are multiple versions, should we just error for now? If there are multiple commits that beat our candidate, it seems like we should generate a WinningCommitSummary for each transaction after `read_version`, right? And then checking our candidate commit against each of them?

Reply:
Good point. Changed the logic, that out commit loop will always only check against the currently conflicting commit and as such the `WinningCommitSummary` will now summarize exactly one commit.